### PR TITLE
Atavedata

### PR DESCRIPTION
**Pull request by ZeusMarti**
The file has been changed to:

1. Avoid large errors on small gradient focusing elements (like quad trim coils in sextupoles).
2. Include orbit generated feed-down effects (orbit at sextupoles).
3. Include rotation and displacement effects on the focusing strength (displaced sextupoles or rotated quadrupoles).
4. Include dispersion formula for general combined function magnets (Formula from A.W.Chao Handbook of Accelerator Physics and Engineering Ed.2002, The Wiedemann formula is wrong).

This tries to solve issue #399

### CODE REVIEW

**Comment by swhite2401** [comment]
Hello @ZeusMarti , this looks nice should we add this to the python too?

**Comment by ZeusMarti** [comment]
Hi @swhite2401, sure, I can take care of it but I'm not python fluent and we are quite busy with alba-II, it can take some time.  Don't take me wrong, I would be happy, for me it would be good a way to force myself to get more familiar with pyat. If you don't feel like waiting for me to go through the exercise please feel free to add the changes yourself. 
Let me know!

**Comment by swhite2401** [comment]
No rush! If you want to train yourself it is very good! Let me know in case you would need any help!

**Comment by ZeusMarti** [comment]
I tested the two implementations in Matlab and Python. Please take a look at it, it should be ready to be reviewed.

Python avllinopt:

Python_avlinopt

Matlab atavedata:
matlab_atavedata_test

**Comment by lfarv** [comment]
Results look nice. Concerning the code, please look at the review above !

**Comment by ZeusMarti** [comment]
> Results look nice. Concerning the code, please look at the review above !

Sorry Laurent, I do not see the review!

**Comment by oscarxblanco** [comment]
Dear @lfarv , neither I see your reviews or requests.
You might need to finish the review to make it public (something similar happened to me while reviewing a few weeks ago)

**Comment by lfarv** [comment]
Hi @ZeusMarti and @oscarxblanco. Is it better now? I was confused because for me it appears in the conversation… Sorry.

File: atmat/atphysics/atavedata.m [hunk 1]
```diff
@@ -40,44 +43,76 @@
...
+        avebeta=beta0-alpha0.*L+gamma0.*L.*L/3;
     end
 
-    function avebeta=betafoc(beta1,alpha0,alpha1,K,L)
-        gamma1=(alpha1.*alpha1+1)./beta1;
-        avebeta=0.5*((gamma1+K.*beta1).*L+alpha1-alpha0)./K./L;
+    function avebeta=betafoc(beta0,alpha0,K,L)
+        gamma0=(alpha0.*alpha0+1)./beta0;
+        avebeta=((beta0+gamma0./K).*L+(beta0-gamma0./K).*sin(2.0*sqrt(K).*L)./sqrt(K)/2.0+(cos(2.0*sqrt(K).*L)-1.0).*alpha0./K)./L/2.0;
     end
```
[the remark this answers is not in the document]
Comment (ZeusMarti) [reply atmat/atphysics/atavedata.m]: See issue [399](https://github.com/atcollab/at/issues/399), using final values diverges for small K values.

File: atmat/atphysics/atavedata.m [hunk 2]
```diff
@@ -40,44 +43,76 @@
...
-    function avedisp=dispfoc(dispp0,dispp1,K,L)
-        avedisp=(dispp0-dispp1)./K./L;
+    function avedisp=dispfoc(disp0,ir,K,L)
+        avedisp=disp0;
+        avedisp(:,[1 3])=(disp0(:,[1 3]).*(sin(sqrt(K).*L)./sqrt(K))+...
+            disp0(:,[2 4]).*(1-cos(sqrt(K).*L))./K)./L+...
+            ir.*(L-sin(sqrt(K).*L)./sqrt(K))./K./L;
+        avedisp(:,[2 4])=(disp0(:,[2 3]).*(sin(sqrt(K).*L)./sqrt(K))-...
+            disp0(:,[1 3]).*(1-cos(sqrt(K).*L)))./L+...
+            ir.*(L-cos(sqrt(K).*L))./K./L;
```
[the remark this answers is not in the document]
Comment (ZeusMarti) [reply atmat/atphysics/atavedata.m:114]: This is new to this atavedata version. To obtained I initially use the Wiedemann book which is very wrong, this one is based on A.W.Chao's Handbook.

File: atmat/atphysics/atavedata.m [hunk 3]
```diff
@@ -40,44 +43,74 @@
...
+        avebeta=beta0-alpha0.*L+gamma0.*L.*L/3;
     end
 
-    function avebeta=betafoc(beta1,alpha0,alpha1,K,L)
-        gamma1=(alpha1.*alpha1+1)./beta1;
-        avebeta=0.5*((gamma1+K.*beta1).*L+alpha1-alpha0)./K./L;
+    function avebeta=betafoc(beta0,alpha0,K,L)
+        gamma0=(alpha0.*alpha0+1)./beta0;
+        avebeta=((beta0+gamma0./K).*L+(beta0-gamma0./K).*sin(2.0*sqrt(K).*L)./sqrt(K)/2.0+...
+            (cos(2.0*sqrt(K).*L)-1.0).*alpha0./K)./L/2.0;
```
[the remark this answers is not in the document]
Comment (ZeusMarti) [reply atmat/atphysics/atavedata.m:104]: See issue [399](https://github.com/atcollab/at/issues/399), using final values diverges for small K values.

File: pyat/at/physics/linear.py [hunk 4]
```diff
@@ -1144,28 +1144,113 @@ def get_strength(elem):
...
-    strength = numpy.array([get_strength(el) for el in ring[boolrefs]])
+    L = numpy.array([el.Length for el in ring[boolrefs]])
+    K = numpy.array([get_strength(el) for el in ring[boolrefs]])
+    sext_strength = numpy.array([get_sext_strength(el) for el in ring[boolrefs]])
+    roll = numpy.array([get_roll(el) for el in ring[boolrefs]])
+    ba = numpy.array([get_bendingangle(el) for el in ring[boolrefs]])
+    e1 = numpy.array([get_e1(el) for el in ring[boolrefs]])
+    Fint = numpy.array([get_fint(el) for el in ring[boolrefs]])
+    gap = numpy.array([get_gap(el) for el in ring[boolrefs]])
+    dx = numpy.array([get_dx(el) for el in ring[boolrefs]])
```
[the remark this answers is not in the document]
Comment (ZeusMarti) [reply pyat/at/physics/linear.py]: Agreed.

**Comment by oscarxblanco** [comment]
Deaer @lfarv , yes. Now the review is visible to me.

File: pyat/at/physics/linear.py [hunk 5]
```diff
@@ -1175,35 +1261,38 @@ def dispfoc(dispp0, dispp1, k2, lg):
     avemu = lindata.mu.copy()
     avedisp = lindata.dispersion.copy()
     aves = lindata.s_pos.copy()
-
+    ClosedOrbit=lindata.closed_orbit.copy()
+
```
Comment (lfarv) [inline pyat/at/physics/linear.py:1265]: `ClosedOrbit` is not used

**Comment by lfarv** [comment]
@ZeusMarti : I get more that 100 warnings about code not respecting PEP8. Most of them are:
```
PEP 8: E225 missing whitespace around operator
PEP 8: E303 too many blank lines (3)
PEP 8: E231 missing whitespace after ','
PEP 8: E128 continuation line under-indented for visual indent
PEP 8: E501 line too long (106 > 88 characters)
```

You should use an IDE which warns about that (ex. PyCharm or VSCode)

**Comment by ZeusMarti** [comment]
@lfarv , I implemented your last comments. 

I realized that I have not properly tested it adding magnet movements, rolls and closed orbit errors, I'll do it after holidays. 

Also, the code logic is a bit cumbersome, there are different behaviors for thin or long, focusing or not, drifts. Any suggestions on how to make it more clear? Maybe it just needs more comments?

**Comment by lfarv** [comment]
>Also, the code logic is a bit cumbersome...

Agreed, but since you are at the moment the most familiar with this code, go on… My only concern is about the "epsilon" trick for the focusing strength. Is not it possible to correctly treat the 0-strength quadrupoles?

> I realized that I have not properly tested it adding magnet movements, rolls and closed orbit errors, I'll do it after holidays.

Excellent !

**Comment by ZeusMarti** [comment]
Hi @lfarv,

I'm sorry I have not been able to advance much, however I noticed a peculiar behaviour of the function, both in Matlab and python in the master branch and this branch. Since the input parameter refpts is used only to create a bolean array, if refpts is not sorted that information is lost in the calculation and the results are given always assuming a sorted order. 

It was the cause of some trouble for me so I'm wondering if this behavior should be kept. We could simply unsort the arrays at the end or alternatively generate a warning in case of non sorted refpts and output the sorted refpts. Which is your opinion on this?

**Comment by lfarv** [comment]
Hi @ZeusMarti : the `refpts` argument, if integer, should always be ordered, and the behaviour you get should be similar in many functions. This is specified in the documentation:
```
**Selecting elements in a lattice:**

The *refpts* argument allows functions to select locations in the lattice. The
location is defined as the entrance of the selected element. *refpts* may be:

#. an integer in the range *[-len(lattice), len(lattice)-1]*
   selecting an element according to the python indexing rules.
   As a special case, *len(lattice)* is allowed and refers to the end
   of the last element,
#. an ordered list of such integers without duplicates,
#. a numpy array of booleans of maximum length *len(lattice)+1*,
   where selected elements are :py:obj:`True`,
...
```
The problem is that it would be difficult (and time consuming) to check the validity of `refpts` everywhere…

Now, for this PR, it's up to you !

**Comment by ZeusMarti** [comment]
Ah, sorry! My bad. Then I see no point in implementing a different behavior for atavedata.m. hopefully I can address the final details of this PR shortly.

**Comment by ZeusMarti** [comment]
Sorry it is taking me so long, we have been very busy lately, I tested the scrips again with errors (up to 100um in x-y and 100urad rotation) and the averages hold very well both in Matlab and Python. I think it is ready fore some more review from your side. 
Matlab:
Python_avlinopt_errors
Python:
matlab_atavedata_test_errors

**Review by lfarv — APPROVED** [review]
That looks very good! Any comment, @swhite2401?

**Review by swhite2401 — APPROVED** [review]
All good for me

**Comment by lfarv** [comment]
Everyone looked happy with this, so I merge

**Pull request merged**